src/encoding_transcode.rs

```diff
@@ -0,0 +1,432 @@
+
+        loop {
+            let (result, read, written, had_errors) = self.inner.encode_from_utf8(
+                &input[total_read_from_input..],
+                &mut self.buffer,
+                false, // not last (since we are processing a continous stream)
+            );
+
+            total_read_from_input += read;
+            total_had_errors |= had_errors;
+
+            self.output.extend_from_slice(&self.buffer[..written]);
+
+            match result {
+                CoderResult::InputEmpty => {
+                    // we have consumed all of the given input so we are done!
+                    break;
+                }
+                CoderResult::OutputFull => {
+                    continue;
+                }
+            }
+        }
+
+        if total_had_errors {
+            warn!(
+                message = "Replaced unmappable characters with numeric character references while encoding from utf8.",
+                to_encoding = %self.inner.encoding().name()
+            );
+        }
+
+        self.output.split().freeze()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{Decoder, Encoder, BOM_UTF8};
+    use bytes::Bytes;
+    use encoding_rs::{SHIFT_JIS, UTF_16BE, UTF_16LE, UTF_8};
+    use std::char::REPLACEMENT_CHARACTER;
+
+    // BOM unicode character (U+FEFF) expressed in utf-16
+    // http://unicode.org/faq/utf_bom.html#bom4
+    const BOM_UTF16LE: &[u8] = b"\xff\xfe";
+
+    // test UTF_16LE data
+    fn test_data_utf16le_123() -> &'static [u8] {
+        b"1\02\03\0"
+    }
+    fn test_data_utf16le_crlf() -> &'static [u8] {
+        b"\r\0\n\0"
+    }
+    fn test_data_utf16le_vector_devanagari() -> &'static [u8] {
+        b"-\tG\t\x15\tM\t\x1f\t0\t"
+    }
+
+    // test UTF_16BE data
+    fn test_data_utf16be_123() -> &'static [u8] {
+        b"\01\02\03"
+    }
+    fn test_data_utf16be_crlf() -> &'static [u8] {
+        b"\0\r\0\n"
+    }
+    fn test_data_utf16be_vector_devanagari() -> &'static [u8] {
+        b"\t-\tG\t\x15\tM\t\x1f\t0"
+    }
+
+    // test SHIFT_JIS data
+    fn test_data_shiftjis_helloworld_japanese() -> &'static [u8] {
+        b"\x83n\x83\x8D\x81[\x81E\x83\x8F\x81[\x83\x8B\x83h"
+    }
+
+    #[test]
+    fn test_decoder_various() {
+        let mut d = Decoder::new(UTF_8);
+        assert_eq!(d.decode_to_utf8(Bytes::from("123")), Bytes::from("123"));
+        assert_eq!(d.decode_to_utf8(Bytes::from("\n")), Bytes::from("\n"));
+        assert_eq!(d.decode_to_utf8(Bytes::from("भेक्टर")), Bytes::from("भेक्टर"));
+
+        let mut d = Decoder::new(UTF_16LE);
+        assert_eq!(
+            d.decode_to_utf8(Bytes::from(test_data_utf16le_123())),
+            Bytes::from("123")
+        );
+        assert_eq!(
+            d.decode_to_utf8(Bytes::from(test_data_utf16le_crlf())),
+            Bytes::from("\r\n")
+        );
+        assert_eq!(
+            d.decode_to_utf8(Bytes::from(test_data_utf16le_vector_devanagari())),
+            Bytes::from("भेक्टर")
```

Comment:
Vector in Devanagari, cool!

Reply:
I don't get to write in it much these days so making the most of it here :smile: